pkg/trace/api/telemetry.go

```diff
@@ -77,32 +87,60 @@ type TelemetryForwarder struct {
 func NewTelemetryForwarder(conf *config.AgentConfig, containerIDProvider IDProvider, statsd statsd.ClientInterface) *TelemetryForwarder {
 	// extract and validate Hostnames from configured endpoints
 	var endpoints []*config.Endpoint
-	for _, endpoint := range conf.TelemetryConfig.Endpoints {
-		u, err := url.Parse(endpoint.Host)
-		if err != nil {
-			log.Errorf("Error parsing apm_config.telemetry endpoint %q: %v", endpoint.Host, err)
-			continue
-		}
-		if u.Host != "" {
-			endpoint.Host = u.Host
-		}
+	if conf.TelemetryConfig != nil {
+		for _, endpoint := range conf.TelemetryConfig.Endpoints {
+			u, err := url.Parse(endpoint.Host)
+			if err != nil {
+				log.Errorf("Error parsing apm_config.telemetry endpoint %q: %v", endpoint.Host, err)
+				continue
+			}
+			if u.Host != "" {
+				endpoint.Host = u.Host
+			}
 
-		endpoints = append(endpoints, endpoint)
+			endpoints = append(endpoints, endpoint)
+		}
 	}
 
-	return &TelemetryForwarder{
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+
+	forwarder := &TelemetryForwarder{
 		endpoints: endpoints,
 		conf:      conf,
 
+		forwardedReqChan: make(chan forwardedRequest),
 		inflightWaiter:   sync.WaitGroup{},
 		inflightCount:    atomic.Int64{},
 		maxInflightBytes: maxInflightBytes,
 
+		cancelCtx: cancelCtx,
+		cancelFn:  cancelFn,
+		done:      make(chan struct{}),
+
 		containerIDProvider: containerIDProvider,
 		client:              conf.NewHTTPClient(),
 		statsd:              statsd,
 		logger:              log.NewThrottled(5, 10*time.Second),
 	}
+	forwarder.start()
+	return forwarder
+}
+
+func (f *TelemetryForwarder) start() {
+	for i := 0; i < maxInflightRequests; i++ {
+		f.inflightWaiter.Add(1)
+		go func() {
```

Comment:
The benchmark on top says between  1.602MB and 1.832MB overhead.
I could reduce the number to something like 100 concurrent requests, but I think we should spawn the goroutines at the start because if we reach the limit the memory is going to be used anyway.

Reply:
there's already one goroutine by http request, why not just quick respond + check if it should be rate limited and if not run the request?

Reply:
> we'd probably be I/O bound anyway and wouldn't benefit from the additional goroutines.

Honestly, on the backend we have kind of a similar setup with currently 400 concurrent goroutines posting http requests to the evp and it works without breaking a sweat, so IO limitations are in practice not really a concern.

And I doubt a few hundreds goroutines doing nothing most of the time puts any real stress most of the time.

Reply:
> And I doubt a few hundreds goroutines doing nothing most of the time puts any real stress most of the time.

While this may be true, we do want to avoid growing the baseline memory usage of the trace-agent overtime and this seems like a pretty straightforward place where we can avoid adding an extra MB of memory. (For context, an idle trace-agent uses about 25Mb currently, this would reflect a 4% increase)

Reply:
Makes sense. I reduced the number of concurrent goroutines to 20, with a bit a buffering (up to 100 inflight requests) because I was afraid of dropping messages in case a lot of apps restart simultaneously and we receive a thundering herd.